Week4/HW04.hs

```diff
@@ -0,0 +1,149 @@
+    where 
+        leadingPlus         = case c of 
+                0           -> "" 
+                _           -> " + "
+        leadingNum          = case p of 
+                1           -> ""
+                anyP        -> show anyP
+
+
+showHelper (P []) _ _       = ""
+
+instance (Num a, Eq a, Show a) => Show (Poly a) where
+    show (P p) = 
+        showHelper (P $ reverse withoutTrailingZeros ) (length withoutTrailingZeros - 1) 0
+            where 
+                withoutTrailingZeros = dropWhileEnd (==0) p
+
+-- Exercise 4 -----------------------------------------
+
+plus :: Num a               => Poly a -> Poly a -> Poly a
+plus (P polyA) (P polyB)    = P $ zipWith (+) (polyA ++ replicate leadingZerosToA 0) (polyB ++ replicate leadingZerosToB 0)
+    where 
+        lA                  = length polyA 
+        lB                  = length polyB
+        leadingZerosToA 
+            | lA < lB       = lB - lA
+            | otherwise     = 0
+        leadingZerosToB 
+            | lB < lA       = lA - lB
+            | otherwise     = 0
+
+-- Exercise 5 -----------------------------------------
+
+-- c is the coeff being multiplied from the first polynomial
+-- e is the exponentioal term being used for multiplication from first polynomial
+
+-- timesHelper (P [1, 4, 2]) 1 0 == P [1, 4, 2]
+-- timesHelper (P [1, 4, 2]) 1 1 == P [0, 1, 4, 2]
+-- timesHelper (P [1, 4, 2]) 1 2 == P [0, 0, 1, 4, 2]
+-- timesHelper (P [1, 4, 2]) 2 2 == P [0, 0, 2, 8, 4]
+
+-- 1. add leading zeros list of size == e 
+-- and multiply rest with c
+timesHelper ::  (Num a)     => Poly a -> a -> Int -> Poly a
+timesHelper (P p) c e       = P $ replicate e 0 ++ map (*c) p
+
+-- getAllTimes 0 (P [1, 1, 2]) (P [1, 4, 2]) == [P [1, 4, 2], P [0, 1, 4, 2], P [0, 0, 2, 8, 4]]
+getAllTimes ::  (Num a)     => Int -> Poly a -> Poly a -> [Poly a]
+getAllTimes _ (P []) _      = []
+getAllTimes n (P [p]) q     = [timesHelper q p n]
+getAllTimes n (P (p:ps)) q  = timesHelper q p n :  getAllTimes (n+1) (P ps) q
+
+-- addAllTimes [P [1, 4, 2], P [0, 1, 4, 2], P [0, 0, 2, 8, 4]] == P [1, 5, 8, 10, 4]
+-- This is how it works with the help of [timesHelper] function
+--      [1, 4, 2, 0, 0]
+--   +  [0, 1, 4, 2, 0]
+--   +  [0, 0, 2, 8, 4]
+--      ---------------
+--      [1, 5, 8, 10, 4]
+--      ---------------
+addAllTimes :: (Num a)      => [Poly a] -> Poly a
+addAllTimes []              = P []
+addAllTimes [p1]            = p1
+addAllTimes [p1, p2]        = plus p1 p2
+addAllTimes (p1:p2:ps)      = plus p1 p2 + addAllTimes ps
+
+times :: (Num a)            => Poly a -> Poly a -> Poly a
+times p q                   = addAllTimes $ getAllTimes 0 p q
+
+-- Exercise 6 -----------------------------------------
+
+instance Num a             => Num (Poly a) where
+    (+)                    = plus
+    (*)                    = times
+    negate                 = (* P [-1])
```

Comment:
While this should work fine, you can do better by just negating all coefficients from the input Poly.

Reply:
You are saying about 2nd right ?

1st
> negate                 = (* P [-1])

2nd
> negate  (P a)           = P $ map negate a

Reply:
Exactly.